src/libraries/System.Private.CoreLib/src/System/ValueTuple.cs

```diff
@@ -516,12 +510,13 @@ int IComparable.CompareTo(object? other)
         {
             if (other == null) return 1;
 
-            if (!(other is ValueTuple<T1, T2>))
+            if (other is not ValueTuple<T1, T2> tuple)
             {
                 ThrowHelper.ThrowArgumentException_TupleIncorrectType(this);
+                tuple = default; // work around CS0170: Use of possibly unassigned field
```

Comment:
What does this do to asm size?  These types get stamped in a lot of places.

Reply:
The JIT is smart enough to recognize the "never returning method", so it won't emit any code for the zeroing of tuple.

If we assume that the case `other is VT` is common, the method should be layed out like

```c#
if (other is ValueTuple<T1> objTuple)
{
    return Comparer<T1>.Default.Compare(Item1, objTuple.Item1);
}

if (other is null) return 1;

ThrowHelper.ThrowArgumentException_TupleIncorrectType(this);
throw null;    // is there any better solution for this?
```

to have this common path on top in the asm. And the call to the comparer, is emitted as tail-call.

One could use pattern matching in a more excessive way too

```c#
return other switch
{
    ValueTuple<T1> objTuple => Comparer<T1>.Default.Compare(Item1, objTuple.Item1),
    null => 1,
    _ => ThrowHelper.ThrowArgumentException_TupleIncorrectType(this)
};
```

(Note: the throw-helper is modified to return `int` to make Roslyn happy).

The codegen for this has a `call` in it, not a tail-call as the suggestion has. So I'd avoid this switch.

[Sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8lAbhvoYGUALbKAA4AZbMAB0AJQCuAOwwBLfDHbUaAbQBS8jAHEYMmFHlgAFBgCeAmBABmJgCYQpwADYwAlO4C6HAMwNcDCgpMAwGADVsFykYABUpATcAHljyAD4aAG8aBlymf1SGAEkMGHxyFRy84n95OQYAYQh8AX44iBMIYAArGFCGCAweQ3cq3OzqPKmGeRsGTqHDBgBeZYYZKRcXdyYAdgYKmjHp2fmAQgXhqBncCKiY+MSYFPSBnse3T2Opiem/hliPCgEAA7gAJGAuKxQMSA4EggCCUAA5lIlHIAKIIMAwAQKCAyAD6HxgRRkkCgsFCsUsMDMPHkuHcKn+0y63RJKwY9hgNmwmwwrAYAHphQwQdAANYMfhOGT2RpcOjkXZ0EAMACquBgAzmAgguFw8lc5gYsjwRuRBgVNnkkPs3zyAF8jpN/sR9k0Wm0oC80mIACK8/kuDBiL2tWAmEplchoN4chJuMQx8rM44u6iZ9RaXT6QzGMy02wOJyuDzePwBIIhMKRaJxJMwBF+rLHGoA8jFUrlSpu3IduphCNtWIddl9MKDK6jfsMX7/U6XJaMu4Nkkt1JpBMk2eshesgee5qRwx+wPBgXhk9taM9uM7psp+/pudTbNvvJL6cr24bLY7B6Bx9qycKghCUKGLCQKgkiqLohgWI4ni8gEsSTZkhSVIYDSVj0oyr6skM8LrJsLgss6NAfpo2h6AYRimBYVglo4zifJWtD+IEwT9PWDxNgAQq21AHvknbdrGIHVLU9QjrAY6dD0k4DIsUB7nkolTEBP7XLgILaGAPCOuMxnTHxjZPEJW6Pk8KzbnJZ5bhefJXg5d6xvG7Iks+sbuGgplTP+Lh2Qc/mfn8hIhWB4KQtC0HwnBaL6Ih2K4viRIkph0DYbhzb4UyplOhRuSZh+Q6GDIURMOQSBMKQAIwTFkFQG2c4dswtXECgDUJSiSWYqlKFoZl5LZZOuWKb0/TsjsyzbsRoLrDAIIMEhaWoTIJiEdJ1W1UOPWwX1CFrUNGUYaNlLjbSCKTcpM0hQtK0GCtJ3pVtKhOkAA==)

Reply:
It results in an additional `call` instruction being emitted. Replacing the assignment with `return 0;` seems to result in nearly identical codegen.

Reply:
Thoughts? If it's not satisfactory I can just revert that particular method to its original form.

Reply:
What about `return ThrowArgumentException_TupleIncorrectType(this)` and change the signature to return int

Reply:
That might constrain us in the hypothetical scenario that this particular throw helper needs to be used with a different return type. Also, it doesn't seem to improve [asm size](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8lAbhvoYGUALbKAA4AZbMAB0AJQCuAOwwBLfDHbUaAbQBS8jAHEYMmFHlgAFBgCeAmBABmJgCYQpwADYwAlO4C6HAMwNcDCgpMAwGADVsFykYABUpATcAHljyAD4aAG8aBlymf1SGAEkMGHxyFRy84n95OQYAYQh8AX44iBMIYAArGFCGCAweQ3cq3OzqPKmGeRsGTqHDBgBeZYYZKRcXdyYAdgYKmjHp2fmAQgXhqBncCKiY+MSYFPSBnse3T2Opiem/hliPCgEAA7gAJGAuKxQMSA4EggCCUAA5lIlHIAKIIMAwAQKCAyAD6HxgRRkkCgsFCsUsMDMPHkuHcKn+02I+0Ok3+AF8jly/uzGs1WrAoC80mIACIwGzYTYYMRNFptEwlMrkNBvboksRq8rM4686hG9RaXT6QzGMy02wOJyuDzePwBIIhMKRaJxBJuBHirLHGoA8jFUrlSr8/IzepKkXtTo9PphQZXUYR37/U6XJardabbZ7A7h1mZi7JpaMu6eknirUkr4Rn7ff6CuGgiFQwywoGgpGo9EYLE4vHyAnE72k8nQKkYGlWBH0xkGht5E3L3KCmNtMWpCXS2XyxXClV6jW18e60PkJdTI2ruqlKAyKJMchIJikAHd8GQ6H+iOB5g32IFBP3hXs0X0AdsVxfEiRJMkKWnWc6S6Xp+lQnZljSBghnhdYYBBBhBxgkcZBMa9qn8QCozCVtERRCDMWg4dR3gydKUTZD51QxMtUw7DcNBfDCOIliyKXbkgA===)

Reply:
Thanks for checking. It's actually worse because RuyJit does not optimize `ret+throw` and emits unnecessary epilog whereas `throw+ret 0` is optimized